File: backend/src/ITAdmin.Application/Common/Security/PermissionCodes.cs
namespace ITAdmin.Application.Common.Security;

public static class PermissionCodes
{
    public static class Dashboard
    {
        public const string View = "Dashboard.View";
    }

    public static class Users
    {
        public const string View = "Users.View";
        public const string Create = "Users.Create";
        public const string Update = "Users.Update";
        public const string Delete = "Users.Delete";
        public const string AssignRoles = "Users.AssignRoles";
    }

    public static class Roles
    {
        public const string View = "Roles.View";
        public const string Create = "Roles.Create";
        public const string Update = "Roles.Update";
        public const string Delete = "Roles.Delete";
        public const string AssignPermissions = "Roles.AssignPermissions";
    }

    public static class Permissions
    {
        public const string View = "Permissions.View";
    }

    public static class AuditLogs
    {
        public const string View = "AuditLogs.View";
    }

    public static class SecurityLogs
    {
        public const string View = "SecurityLogs.View";
    }

    public static class Settings
    {
        public const string View = "Settings.View";
        public const string Update = "Settings.Update";
    }

    public static class Setup
    {
        public const string Manage = "Setup.Manage";
    }

    public static class NotificationProviders
    {
        public const string View = "NotificationProviders.View";
        public const string Update = "NotificationProviders.Update";
        public const string Test = "NotificationProviders.Test";
    }

    public static class NotificationOutbox
    {
        public const string View = "NotificationOutbox.View";
        public const string Retry = "NotificationOutbox.Retry";
        public const string Cancel = "NotificationOutbox.Cancel";
    }

    public static class NotificationTemplates
    {
        public const string View = "NotificationTemplates.View";
        public const string Update = "NotificationTemplates.Update";
    }

    public static class AdOperationLogs
    {
        public const string View = "AdOperationLogs.View";
    }

    public static class AdManagement
    {
        public static class Settings
        {
            public const string View = "AdManagement.Settings.View";
            public const string Update = "AdManagement.Settings.Update";
        }

        public static class Users
        {
            public const string View = "AdManagement.Users.View";
            public const string Create = "AdManagement.Users.Create";
            public const string Update = "AdManagement.Users.Update";
            public const string Enable = "AdManagement.Users.Enable";
            public const string Disable = "AdManagement.Users.Disable";
            public const string Unlock = "AdManagement.Users.Unlock";
            public const string MoveOu = "AdManagement.Users.MoveOu";

            public static class Groups
            {
                public const string View = "AdManagement.Users.Groups.View";
                public const string Add = "AdManagement.Users.Groups.Add";
                public const string Remove = "AdManagement.Users.Groups.Remove";
            }
        }

        public static class Groups
        {
            public const string View = "AdManagement.Groups.View";
            public const string Create = "AdManagement.Groups.Create";
            public const string Update = "AdManagement.Groups.Update";
            public const string Delete = "AdManagement.Groups.Delete";
            public const string ManageMembers = "AdManagement.Groups.ManageMembers";
            public const string MoveOu = "AdManagement.Groups.MoveOu";
        }

        public static class Computers
        {
            public const string View = "AdManagement.Computers.View";
            public const string Update = "AdManagement.Computers.Update";
            public const string MoveOu = "AdManagement.Computers.MoveOu";
            public const string Enable = "AdManagement.Computers.Enable";
            public const string Disable = "AdManagement.Computers.Disable";
            public const string Delete = "AdManagement.Computers.Delete";

            public static class Groups
            {
                public const string View = "AdManagement.Computers.Groups.View";
                public const string Add = "AdManagement.Computers.Groups.Add";
                public const string Remove = "AdManagement.Computers.Groups.Remove";
            }
        }

        public static class DeletedObjects
        {
            public const string View = "AdManagement.DeletedObjects.View";
            public const string Restore = "AdManagement.DeletedObjects.Restore";
        }

        public static class OrganizationalUnits
        {
            public const string View = "AdManagement.OrganizationalUnits.View";
            public const string Create = "AdManagement.OrganizationalUnits.Create";
            public const string Update = "AdManagement.OrganizationalUnits.Update";
            public const string Move = "AdManagement.OrganizationalUnits.Move";
            public const string Delete = "AdManagement.OrganizationalUnits.Delete";
        }
    }

    public static class LicenseManagement
    {
        public const string View = "LicenseManagement.View";
        public const string ManageCatalog = "LicenseManagement.ManageCatalog";
        public const string ManagePurchases = "LicenseManagement.ManagePurchases";
        public const string ManageRequests = "LicenseManagement.ManageRequests";
        public const string FulfillRequests = "LicenseManagement.FulfillRequests";
        public const string ViewReports = "LicenseManagement.ViewReports";
        public const string ManageSettings = "LicenseManagement.ManageSettings";
    }

    public static class Directory
    {
        public static class Users
        {
            public const string Lookup = "Directory.Users.Lookup";
        }

        public static class OrganizationalUnits
        {
            public const string Lookup = "Directory.OrganizationalUnits.Lookup";
        }
    }
}

File: frontend/src/lib/permission-codes.ts
export const PermissionCodes = {
  Dashboard: {
    View: "Dashboard.View",
  },
  Users: {
    View: "Users.View",
    Create: "Users.Create",
    Update: "Users.Update",
    Delete: "Users.Delete",
    AssignRoles: "Users.AssignRoles",
  },
  Roles: {
    View: "Roles.View",
    Create: "Roles.Create",
    Update: "Roles.Update",
    Delete: "Roles.Delete",
    AssignPermissions: "Roles.AssignPermissions",
  },
  Permissions: {
    View: "Permissions.View",
  },
  AuditLogs: {
    View: "AuditLogs.View",
  },
  SecurityLogs: {
    View: "SecurityLogs.View",
  },
  Settings: {
    View: "Settings.View",
    Update: "Settings.Update",
  },
  SystemUpdates: {
    View: "System.Updates.View",
    Manage: "System.Updates.Manage",
  },
  SystemHttps: {
    View: "System.Https.View",
    Manage: "System.Https.Manage",
  },
  Setup: {
    Manage: "Setup.Manage",
  },
  NotificationProviders: {
    View: "NotificationProviders.View",
    Update: "NotificationProviders.Update",
    Test: "NotificationProviders.Test",
  },
  NotificationOutbox: {
    View: "NotificationOutbox.View",
    Retry: "NotificationOutbox.Retry",
    Cancel: "NotificationOutbox.Cancel",
  },
  NotificationTemplates: {
    View: "NotificationTemplates.View",
    Update: "NotificationTemplates.Update",
  },
  AdOperationLogs: {
    View: "AdOperationLogs.View",
  },
  AdManagement: {
    Settings: {
      View: "AdManagement.Settings.View",
      Update: "AdManagement.Settings.Update",
    },
    Users: {
      View: "AdManagement.Users.View",
      Create: "AdManagement.Users.Create",
      Update: "AdManagement.Users.Update",
      Enable: "AdManagement.Users.Enable",
      Disable: "AdManagement.Users.Disable",
      Unlock: "AdManagement.Users.Unlock",
      MoveOu: "AdManagement.Users.MoveOu",
      Groups: {
        View: "AdManagement.Users.Groups.View",
        Add: "AdManagement.Users.Groups.Add",
        Remove: "AdManagement.Users.Groups.Remove",
      },
    },
    Groups: {
      View: "AdManagement.Groups.View",
      Create: "AdManagement.Groups.Create",
      Update: "AdManagement.Groups.Update",
      Delete: "AdManagement.Groups.Delete",
      ManageMembers: "AdManagement.Groups.ManageMembers",
      MoveOu: "AdManagement.Groups.MoveOu",
    },
    Computers: {
      View: "AdManagement.Computers.View",
      Update: "AdManagement.Computers.Update",
      MoveOu: "AdManagement.Computers.MoveOu",
      Enable: "AdManagement.Computers.Enable",
      Disable: "AdManagement.Computers.Disable",
      Delete: "AdManagement.Computers.Delete",
      Groups: {
        View: "AdManagement.Computers.Groups.View",
        Add: "AdManagement.Computers.Groups.Add",
        Remove: "AdManagement.Computers.Groups.Remove",
      },
    },
    DeletedObjects: {
      View: "AdManagement.DeletedObjects.View",
      Restore: "AdManagement.DeletedObjects.Restore",
    },
    OrganizationalUnits: {
      View: "AdManagement.OrganizationalUnits.View",
      Create: "AdManagement.OrganizationalUnits.Create",
      Update: "AdManagement.OrganizationalUnits.Update",
      Move: "AdManagement.OrganizationalUnits.Move",
      Delete: "AdManagement.OrganizationalUnits.Delete",
    },
  },
  LicenseManagement: {
    View: "LicenseManagement.View",
    ViewSensitiveData: "LicenseManagement.ViewSensitiveData",
    ManageCatalog: "LicenseManagement.ManageCatalog",
    ManagePurchases: "LicenseManagement.ManagePurchases",
    ManageRequests: "LicenseManagement.ManageRequests",
    FulfillRequests: "LicenseManagement.FulfillRequests",
    ViewReports: "LicenseManagement.ViewReports",
    ManageSettings: "LicenseManagement.ManageSettings",
  },
  DnsManagement: {
    View: "DnsManagement.View",
    ManageSettings: "DnsManagement.ManageSettings",
    Servers: { View: "DnsManagement.Servers.View", Manage: "DnsManagement.Servers.Manage", TestConnection: "DnsManagement.Servers.TestConnection" },
    Zones: { View: "DnsManagement.Zones.View", Create: "DnsManagement.Zones.Create", Update: "DnsManagement.Zones.Update", Delete: "DnsManagement.Zones.Delete" },
    Records: { View: "DnsManagement.Records.View", Create: "DnsManagement.Records.Create", Update: "DnsManagement.Records.Update", Delete: "DnsManagement.Records.Delete" },
    Compare: "DnsManagement.Compare", Export: "DnsManagement.Export", Synchronize: "DnsManagement.Synchronize",
    ManageServerSettings: "DnsManagement.ServerSettings.Manage", ManageDnssec: "DnsManagement.Dnssec.Manage", ManagePolicies: "DnsManagement.Policies.Manage", ClearCache: "DnsManagement.Cache.Clear", ManageScavenging: "DnsManagement.Scavenging.Manage", ManageNetworkConfiguration: "DnsManagement.NetworkConfiguration.Manage", ViewOperationLogs: "DnsManagement.OperationLogs.View",
  },
  Directory: {
    Users: {
      Lookup: "Directory.Users.Lookup",
    },
    OrganizationalUnits: {
      Lookup: "Directory.OrganizationalUnits.Lookup",
    },
  },
} as const;

type DeepPermissionValue<T> = T extends string
  ? T
  : T extends Record<string, unknown>
    ? DeepPermissionValue<T[keyof T]>
    : never;

export type PermissionCode = DeepPermissionValue<typeof PermissionCodes>;
